server/src/tracker.rs

```diff
@@ -0,0 +1,254 @@
+use std::collections::HashMap;
+use std::future::Future;
+use std::pin::Pin;
+use std::sync::atomic::{AtomicUsize, Ordering};
+use std::sync::{Arc, Mutex};
+use std::task::{Context, Poll};
+
+use pin_project::{pin_project, pinned_drop};
+use tokio::sync::oneshot;
+
+/// Every future registered with a `TrackerContext` is assigned a unique
+/// `TrackerId`
+#[derive(Clone, Copy, Debug, Hash, PartialEq, Eq, PartialOrd, Ord)]
+pub struct TrackerId(usize);
+
+#[derive(Debug)]
+struct Tracker<T> {
+    data: T,
+    signal: Option<oneshot::Sender<()>>,
+}
+
+#[derive(Debug)]
+struct TrackerContextInner<T> {
+    id: AtomicUsize,
+    trackers: Mutex<HashMap<TrackerId, Tracker<T>>>,
+}
+
+/// Allows tracking the lifecycle of futures registered by
+/// `TrackedFutureExt::track` with an accompanying metadata payload of type T
+///
+/// Additionally can trigger graceful termination of registered futures
+#[derive(Debug)]
+pub struct TrackerContext<T> {
+    inner: Arc<TrackerContextInner<T>>,
+}
+
+// Manual Clone to workaround https://github.com/rust-lang/rust/issues/26925
+impl<T> Clone for TrackerContext<T> {
+    fn clone(&self) -> Self {
+        Self {
+            inner: self.inner.clone(),
+        }
+    }
+}
+
+impl<T> Default for TrackerContext<T> {
+    fn default() -> Self {
+        Self {
+            inner: Arc::new(TrackerContextInner {
+                id: AtomicUsize::new(0),
+                trackers: Mutex::new(Default::default()),
+            }),
+        }
+    }
+}
+
+impl<T> TrackerContext<T> {
+    pub fn new() -> Self {
+        Default::default()
+    }
+
+    /// Trigger graceful termination of a registered future
```

Comment:
Is this more accurate:
```suggestion
    /// Trigger graceful termination of a registered future if it has not yet started executing
```

Reply:
It's more subtle than that, the abort kicks in within poll and so if poll is never called or the process never yields (i.e. returns Poll::Pending) it will have no impact. I've updated to what I think might be clearer?